### PR TITLE
Allow sending email without "To" if "Cc" or "Bcc" is provided

### PR DESCRIPTION
**Pull request by hishamco**
Addresses #10840

### CODE REVIEW

File: src/OrchardCore/OrchardCore.Email.Core/Services/SmtpService.cs [hunk 1]
```diff
@@ -34,22 +43,33 @@ IStringLocalizer<SmtpService> stringLocalizer
             S = stringLocalizer;
         }
 
+        /// <summary>
+        /// Sends the specified message to an SMTP server for delivery.
+        /// </summary>
+        /// <param name="message">The message to be sent.</param>
+        /// <returns>A <see cref="SmtpResult"/> that holds information about the sent message, for instance if it has sent successfully or if it has failed.</returns>
+        /// <remarks>This method allows to send an email without setting <see cref="MailMessage.To"/> if <see cref="MailMessage.Cc"/> or <see cref="MailMessage.Bcc"/> is provided.</remarks>
```
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Email.Core/Services/SmtpService.cs:51]: I didn't see a suitable place to mention this in our docs, so I just write here, so any developer can notice that the email can required on of the three headers

**Comment by hishamco** [comment]
I may need to check work flow task too, to react to this change

**Comment by Skrypt** [comment]
Yes, here.

https://github.com/OrchardCMS/OrchardCore/blob/e73a81ee82bb55fad2a4ac5d3d1fe6fd986f28b4/src/OrchardCore.Modules/OrchardCore.Email/Workflows/ViewModels/EmailTaskViewModel.cs#L17-L18

File: src/OrchardCore/OrchardCore.Email.Core/Services/SmtpService.cs [hunk 2]
```diff
@@ -34,22 +43,38 @@ IStringLocalizer<SmtpService> stringLocalizer
...
+                if (!String.IsNullOrWhiteSpace(senderAddress))
+                {
+                    message.From = senderAddress;
+                }
+
                 var mimeMessage = FromMailMessage(message);
 
+                if (mimeMessage.From.Count == 0 && mimeMessage.Cc.Count == 0 && mimeMessage.Bcc.Count == 0)
+                {
+                    return SmtpResult.Failed(S["The mail message should have at least one of these headers: To, Cc or Bcc."]);
```
Comment (Skrypt) [inline src/OrchardCore/OrchardCore.Email.Core/Services/SmtpService.cs:75]: I'm asking myself if the Mailkit library now won't return proper error messages when we update it to the latest version like you saw here https://github.com/OrchardCMS/OrchardCore/pull/10863

Comment (Skrypt) [inline src/OrchardCore/OrchardCore.Email.Core/Services/SmtpService.cs:75]: I'm asking myself if the Mailkit library now won't return proper error messages when we update it to the latest version like you saw here #10863

I would tend to let the message fail without throwing a specific error message here. Let the library do its work if that's possible.

File: test/OrchardCore.Tests/Modules/OrchardCore.Email/Workflows/EmailTaskTests.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+using OrchardCore.Email.Workflows.Activities;
+using OrchardCore.Workflows.Models;
+using OrchardCore.Workflows.Services;
+using Xunit;
+
+namespace OrchardCore.Tests.Modules.OrchardCore.Email.Workflows
+{
+    public class EmailTaskTests
+    {
+        private static readonly IDictionary<string, object> EmptyDictiony = new Dictionary<string, object>();
```
Comment (Skrypt) [inline test/OrchardCore.Tests/Modules/OrchardCore.Email/Workflows/EmailTaskTests.cs:21]: EmptyDictionary

File: test/OrchardCore.Tests/Modules/OrchardCore.Email/Workflows/EmailTaskTests.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                HtmlEncoder.Default)
+            {
+                Subject = new WorkflowExpression<string>("Test"),
+                Body = new WorkflowExpression<string>("Test message!!")
+            };
+
+            var executionContext = new WorkflowExecutionContext(
+                new WorkflowType(),
+                new Workflow(),
+                EmptyDictiony,
```
Comment (Skrypt) [inline test/OrchardCore.Tests/Modules/OrchardCore.Email/Workflows/EmailTaskTests.cs:40]: EmptyDictionary

File: test/OrchardCore.Tests/Modules/OrchardCore.Email/Workflows/EmailTaskTests.cs (hunk 3 above)
Comment (hishamco) [reply test/OrchardCore.Tests/Modules/OrchardCore.Email/Workflows/EmailTaskTests.cs:21]: Good catch

File: src/OrchardCore/OrchardCore.Email.Core/Services/SmtpService.cs (hunk 2 above)
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Email.Core/Services/SmtpService.cs:75]: I think there's no throwing here, I suggest to log an explicit message to make it clear. I'm not sure if the server reponse returns a specific or general message

Let's accept this as its, then we can add the server response to `SmtpResult` as I suggested in related link

**Review by Skrypt — APPROVED** [review]

**Pull request merged**